vue.config.js failing calls to require() are now ignored

### Version
4.3.0

### Reproduction link
[https://github.com/mrozekma/vue-cli-bug-demo](https://github.com/mrozekma/vue-cli-bug-demo)





### Environment info
```
Environment Info:

  System:
    OS: Linux 4.19 Debian GNU/Linux 9 (stretch) 9 (stretch)
    CPU: (4) x64 06/8e
  Binaries:
    Node: 12.10.0 - /usr/bin/node
    Yarn: 1.19.1 - /usr/bin/yarn
    npm: 6.10.3 - /usr/bin/npm
  Browsers:
    Chrome: Not Found
    Firefox: 68.7.0esr
  npmPackages:
    @vue/babel-helper-vue-jsx-merge-props:  1.0.0 
    @vue/babel-plugin-transform-vue-jsx:  1.1.2 
    @vue/babel-preset-app:  4.3.1 
    @vue/babel-preset-jsx:  1.1.2 
    @vue/babel-sugar-functional-vue:  1.1.2 
    @vue/babel-sugar-inject-h:  1.1.2 
    @vue/babel-sugar-v-model:  1.1.2 
    @vue/babel-sugar-v-on:  1.1.2 
    @vue/cli-overlay:  4.3.1 
    @vue/cli-plugin-babel: ^4.3.0 => 4.3.1 
    @vue/cli-plugin-router: ^4.3.0 => 4.3.1 
    @vue/cli-plugin-typescript: ^4.3.0 => 4.3.1 
    @vue/cli-plugin-vuex:  4.3.1 
    @vue/cli-service: ^4.3.0 => 4.3.1 
    @vue/cli-shared-utils:  4.3.1 
    @vue/component-compiler-utils:  3.1.2 
    @vue/preload-webpack-plugin:  1.1.1 
    @vue/web-component-wrapper:  1.2.0 
    typescript: ~3.8.3 => 3.8.3 
    vue: ^2.6.11 => 2.6.11 
    vue-hot-reload-api:  2.3.4 
    vue-loader:  15.9.2 
    vue-router: ^3.1.6 => 3.1.6 
    vue-style-loader:  4.1.2 
    vue-template-compiler: ^2.6.11 => 2.6.11 
    vue-template-es2015-compiler:  1.9.1 
  npmGlobalPackages:
    @vue/cli: 4.2.2

```


### Steps to reproduce
Run `yarn serve`. It will load successfully. Then downgrade to @vue/cli-service@4.2.3 and `yarn serve` will correctly fail.

### What is expected?
 ERROR  Error loading vue.config.js:  
 ERROR  Error: Cannot find module 'bad-module'


### What is actually happening?
Server loads successfully

---
Commit f1bdf7319ac3daa54d57589fc99d7b33696ac0a1 simplified `vue.config.js` loading by not checking if the file exists, but instead just `require()`-ing it and catching the failure if it doesn't. However, this also suppresses cases where `vue.config.js` exists, but it in turn calls `require()` on something that doesn't exist, since all the exception handling code looks for is `e.code === 'MODULE_NOT_FOUND'`

<!-- generated by vue-issues. DO NOT REMOVE -->

--- a/packages/@vue/cli-service/__tests__/Service.spec.js
+++ b/packages/@vue/cli-service/__tests__/Service.spec.js
@@ -130,7 +130,7 @@ test('keep publicPath when empty', () => {
 })
 
 test('load project options from vue.config.js', () => {
-  fs.writeFileSync('/vue.config.js', '')  // only to ensure fs.existsSync returns true
+  fs.writeFileSync(path.resolve('/', 'vue.config.js'), '')  // only to ensure fs.existsSync returns true
   jest.mock(path.resolve('/', 'vue.config.js'), () => ({ lintOnSave: false }), { virtual: true })
   mockPkg({
     vue: {
@@ -143,8 +143,8 @@ test('load project options from vue.config.js', () => {
 })
 
 test('load project options from vue.config.js as a function', () => {
-  fs.writeFileSync('/vue.config.js', '')
-  jest.mock('/vue.config.js', () => function () { return { lintOnSave: false } }, { virtual: true })
+  fs.writeFileSync(path.resolve('/', 'vue.config.js'), '')
+  jest.mock(path.resolve('/', 'vue.config.js'), () => function () { return { lintOnSave: false } }, { virtual: true })
   mockPkg({
     vue: {
       lintOnSave: 'default'

--- a/packages/@vue/cli-service/__tests__/Service.spec.js
+++ b/packages/@vue/cli-service/__tests__/Service.spec.js
@@ -29,6 +29,12 @@ beforeEach(() => {
   delete process.env.BAZ
 })
 
+afterEach(() => {
+  if (fs.existsSync('/vue.config.js')) {
+    fs.unlinkSync('/vue.config.js')
+  }
+})
+
 test('env loading', () => {
   process.env.FOO = 0
   fs.writeFileSync('/.env.local', `FOO=1\nBAR=2`)
@@ -124,6 +130,7 @@ test('keep publicPath when empty', () => {
 })
 
 test('load project options from vue.config.js', () => {
+  fs.writeFileSync('/vue.config.js', '')  // only to ensure fs.existsSync returns true
   jest.mock(path.resolve('/', 'vue.config.js'), () => ({ lintOnSave: false }), { virtual: true })
   mockPkg({
     vue: {
@@ -136,6 +143,7 @@ test('load project options from vue.config.js', () => {
 })
 
 test('load project options from vue.config.js as a function', () => {
+  fs.writeFileSync('/vue.config.js', '')
   jest.mock('/vue.config.js', () => function () { return { lintOnSave: false } }, { virtual: true })
   mockPkg({
     vue: {

--- a/packages/@vue/cli-service/__tests__/ServiceESM.spec.js
+++ b/packages/@vue/cli-service/__tests__/ServiceESM.spec.js
@@ -1,12 +1,13 @@
 const Service = require('../lib/Service')
 
-const configPath = '/vue.config.cjs'
+const path = require('path')
+const configPath = path.resolve('/', 'vue.config.cjs')
 
 jest.mock('fs')
 const fs = require('fs')
 
 beforeEach(() => {
-  fs.writeFileSync('/package.json', JSON.stringify({
+  fs.writeFileSync(path.resolve('/', 'package.json'), JSON.stringify({
     type: 'module',
     vue: {
       lintOnSave: 'default'

--- a/packages/@vue/cli-service/__tests__/ServiceESM.spec.js
+++ b/packages/@vue/cli-service/__tests__/ServiceESM.spec.js
@@ -1,11 +1,27 @@
-const { join } = require('path')
 const Service = require('../lib/Service')
 
-const mockDir = join(__dirname, 'mockESM')
-const configPath = join(mockDir, 'vue.config.cjs')
+const configPath = '/vue.config.cjs'
+
+jest.mock('fs')
+const fs = require('fs')
+
+beforeEach(() => {
+  fs.writeFileSync('/package.json', JSON.stringify({
+    type: 'module',
+    vue: {
+      lintOnSave: 'default'
+    }
+  }, null, 2))
+})
+
+afterEach(() => {
+  if (fs.existsSync(configPath)) {
+    fs.unlinkSync(configPath)
+  }
+})
 
 const createService = () => {
-  const service = new Service(mockDir, {
+  const service = new Service('/', {
     plugins: [],
     useBuiltIn: false
   })
@@ -21,12 +37,14 @@ test('load project options from package.json', async () => {
 })
 
 test('load project options from vue.config.cjs', async () => {
+  fs.writeFileSync(configPath, '')
   jest.mock(configPath, () => ({ lintOnSave: true }), { virtual: true })
   const service = createService()
   expect(service.projectOptions.lintOnSave).toBe(true)
 })
 
 test('load project options from vue.config.cjs as a function', async () => {
+  fs.writeFileSync(configPath, '')
   jest.mock(configPath, () => function () { return { lintOnSave: true } }, { virtual: true })
   const service = createService()
   expect(service.projectOptions.lintOnSave).toBe(true)

--- a/packages/@vue/cli-service/__tests__/mockESM/package.json
+++ b/packages/@vue/cli-service/__tests__/mockESM/package.json
@@ -1,6 +1,0 @@
-{
-  "type": "module",
-  "vue": {
-    "lintOnSave": "default"
-  }
-}

--- a/packages/@vue/cli-service/lib/Service.js
+++ b/packages/@vue/cli-service/lib/Service.js
@@ -1,3 +1,4 @@
+const fs = require('fs')
 const path = require('path')
 const debug = require('debug')
 const merge = require('webpack-merge')
@@ -9,22 +10,6 @@ const defaultsDeep = require('lodash.defaultsdeep')
 const { chalk, warn, error, isPlugin, resolvePluginId, loadModule, resolvePkg } = require('@vue/cli-shared-utils')
 
 const { defaults, validate } = require('./options')
-
-const loadConfig = configPath => {
-  let fileConfig = require(configPath)
-
-  if (typeof fileConfig === 'function') {
-    fileConfig = fileConfig()
-  }
-
-  if (!fileConfig || typeof fileConfig !== 'object') {
-    error(
-      `Error loading ${chalk.bold('vue.config.js')}: should export an object or a function that returns object.`
-    )
-    fileConfig = null
-  }
-  return fileConfig
-}
 
 module.exports = class Service {
   constructor (context, { plugins, pkg, inlineOptions, useBuiltIn } = {}) {
@@ -314,41 +299,46 @@ module.exports = class Service {
   }
 
   loadUserOptions () {
-    // vue.config.js
-    // vue.config.cjs
+    // vue.config.c?js
     let fileConfig, pkgConfig, resolved, resolvedFrom
     const esm = this.pkg.type && this.pkg.type === 'module'
-    const jsConfigPath = path.resolve(this.context, 'vue.config.js')
-    const cjsConfigPath = path.resolve(this.context, 'vue.config.cjs')
-    const configPath = (
-      process.env.VUE_CLI_SERVICE_CONFIG_PATH ||
-      jsConfigPath
-    )
 
-    try {
-      fileConfig = loadConfig(configPath)
-    } catch (e) {
-      if (e.code !== 'MODULE_NOT_FOUND') {
-        if (e.code === 'ERR_REQUIRE_ESM') {
-          warn(`Rename ${chalk.bold('vue.config.js')} to ${chalk.bold('vue.config.cjs')} when ECMAScript modules is enabled`)
-        }
-        error(`Error loading ${chalk.bold('vue.config.js')}:`)
-        throw e
+    const possibleConfigPaths = [
+      process.env.VUE_CLI_SERVICE_CONFIG_PATH,
+      './vue.config.js',
+      './vue.config.cjs'
+    ]
+
+    let fileConfigPath
+    for (const p of possibleConfigPaths) {
+      const resolvedPath = p && path.resolve(this.context, p)
+      if (resolvedPath && fs.existsSync(resolvedPath)) {
+        fileConfigPath = resolvedPath
       }
     }
 
-    // vue.config.js not found, esm enabled, no env set
-    if (!fileConfig && esm && !process.env.VUE_CLI_SERVICE_CONFIG_PATH) {
-      try {
-        fileConfig = loadConfig(cjsConfigPath)
-      } catch (e) {
-        if (e.code !== 'MODULE_NOT_FOUND') {
-          error(`Error loading ${chalk.bold('vue.config.cjs')}:`)
-          throw e
-        }
+    if (fileConfigPath) {
+      if (esm && fileConfigPath === './vue.config.js') {
+        throw new Error(`Please rename ${chalk.bold('vue.config.js')} to ${chalk.bold('vue.config.cjs')} when ECMAScript modules is enabled`)
       }
-      if (fileConfig) {
-        warn(`ECMAScript modules is detected, config loaded from ${chalk.bold('vue.config.cjs')}`)
+
+      try {
+        fileConfig = loadModule(fileConfigPath, this.context)
+
+        if (typeof fileConfig === 'function') {
+          fileConfig = fileConfig()
+        }
+
+        if (!fileConfig || typeof fileConfig !== 'object') {
+          // TODO: show throw an Error here, to be fixed in v5
+          error(
+            `Error loading ${chalk.bold(fileConfigPath)}: should export an object or a function that returns object.`
+          )
+          fileConfig = null
+        }
+      } catch (e) {
+        error(`Error loading ${chalk.bold(fileConfigPath)}:`)
+        throw e
       }
     }
 

--- a/packages/@vue/cli-shared-utils/lib/module.js
+++ b/packages/@vue/cli-shared-utils/lib/module.js
@@ -61,8 +61,10 @@ exports.resolveModule = function (request, context) {
 }
 
 exports.loadModule = function (request, context, force = false) {
-  // createRequire doesn't work with jest mock modules (which we used in migrator, for inquirer)
-  if (process.env.VUE_CLI_TEST && request.endsWith('migrator')) {
+  // createRequire doesn't work with jest mock modules
+  // (which we used in migrator for inquirer, and in tests for cli-service)
+  // TODO: it's supported in Jest 25
+  if (process.env.VUE_CLI_TEST && (request.endsWith('migrator') || context === '/')) {
     return require(request)
   }
 

--- a/packages/@vue/cli-shared-utils/lib/module.js
+++ b/packages/@vue/cli-shared-utils/lib/module.js
@@ -65,6 +65,9 @@ exports.loadModule = function (request, context, force = false) {
   // (which we used in migrator for inquirer, and in tests for cli-service)
   // TODO: it's supported in Jest 25
   if (process.env.VUE_CLI_TEST && (request.endsWith('migrator') || context === '/')) {
+    if (context === '/') {
+      request = request.replace('C:\\', '/')
+    }
     return require(request)
   }
 

--- a/packages/@vue/cli-shared-utils/lib/module.js
+++ b/packages/@vue/cli-shared-utils/lib/module.js
@@ -65,9 +65,6 @@ exports.loadModule = function (request, context, force = false) {
   // (which we used in migrator for inquirer, and in tests for cli-service)
   // TODO: it's supported in Jest 25
   if (process.env.VUE_CLI_TEST && (request.endsWith('migrator') || context === '/')) {
-    if (context === '/') {
-      request = request.replace('C:\\', '/')
-    }
     return require(request)
   }
 

